### PR TITLE
Add debug cookie option

### PR DESCRIPTION
**Pull request by unlocomqx**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add ability to configure a debug cookie to control the debug mode
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Open the performance page and enable the debug mode, then configure a cookie and a value
| Fixed ticket?     | Discussion https://github.com/PrestaShop/PrestaShop/discussions/30840
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/660
| Sponsor company   | 

We can add other options. I only added the cookie name & value for now. It's more of a simple proof of concept. Tell me if something is too hacky or potentially unreliable.

CleanShot 2023-06-02 at 23 04 19@2x

### CODE REVIEW

File: src/PrestaShopBundle/Form/Admin/AdvancedParameters/Performance/DebugModeType.php [hunk 1]
```diff
@@ -27,7 +27,9 @@
 namespace PrestaShopBundle\Form\Admin\AdvancedParameters\Performance;
 
 use PrestaShopBundle\Form\Admin\Type\SwitchType;
+use PrestaShopBundle\Form\Admin\Type\TextPreviewType;
```
Comment (kpodemski) [inline src/PrestaShopBundle/Form/Admin/AdvancedParameters/Performance/DebugModeType.php]:
```suggestion
```

Comment (unlocomqx) [reply src/PrestaShopBundle/Form/Admin/AdvancedParameters/Performance/DebugModeType.php]: I removed the unused import, thanks

**Comment by MatShir** [comment]
Thank you for initiating this POC. 
I have two questions:

- Is setting a value by default, will not be risky or exploitable?
I wonder if a generated code system like code cart rules would be more practical.

- And I wonder if we shouldn't review the switches? 
People might instinctively activate debug mode before configuring cookies. So, either set a default value, so revisit the first question or if we don’t want a default value, ask before activating the debug mode if the user want to set cookies.

**Comment by unlocomqx** [comment]
Hi @MatShir 

- Settings a default value in this case is like setting a default password. Every attacker will try it first and few users will change it. So it's risky and easily exploitable.
To be clear, I only put a placeholder, not a default value. I can remove the placeholder if you think some merchants will think it's a default value or is supposed to be a default value.
Also, we don't need a default value I think, because most merchants will simply want to enable the debug mode, without a cookie condition (Such as when they are testing locally/under maintenance mode etc)
- I think that the flow is intuitive. First you enable the switch and the fields are shown, prompting the user's attention, so they can optionally fill them if they need and then save the values. It's the same as the "Cache" switch above it.

**Comment by MatShir** [comment]
Hi @unlocomqx 
Thanks for the details. 
No placeholder would be better. The behavior described look promising. 
If at some point, we want to add the profiler and other feature depends on the cookie, we can improve it later. We can imagine a proper panel for the cookie configuration with switch for a general setting or add the switch for each option...

**Comment by jf-viguier** [comment]
Why asking for a cookie and not use the admin psAdmin cookie ?

**Comment by unlocomqx** [comment]
> Why asking for a cookie and not use the admin psAdmin cookie ?

Sometimes, you don't want to enable debug mode for all admins as it may disrupt their work. 

In other cases, you can allow a developer to see the shop in debug mode without giving them admin access.

What do you think?

**Comment by jf-viguier** [comment]
the psAdmin cookie contains current employee is, it's not restricted to super admin account.

> In other cases, you can allow a developer to see the shop in debug mode without giving them admin access.

you can create a simple employee account (not super admin)

**Comment by kpodemski** [comment]
Hi @unlocomqx 

We agreed with @MatShir that your idea is a very good first step and covers most of the scenarios we could think of. The one thing that could be interesting is adding a button to the next of each field you added that helps the merchant generate the values.

I can imagine that it'd work like this:
1. Merchant bought a module, but it doesn't work, error 500
2. He reaches out to support
3. Support asks him to provide these two values so that they can enable the debug mode
4. He generates the values and provides them with the support

wdyt?

**Comment by unlocomqx** [comment]
@kpodemski Thank you. I will see how to add a button to generate random values.

**Comment by jf-viguier** [comment]
@kpodemski is the front is in 500 error, It could be the same in backoffice. So maybe we need an independant tool to activate the debug with all prestashop blocked with a php error ?

**Comment by jf-viguier** [comment]
Nice PR @unlocomqx thanks.

- I don't know why the cookie name is parametrable ? it could be fixed
- The cookie value should be generated to be non guessable, my be 15 c or more
- The cookie value should be time limited
- This user case could be adapted to the profiler and maintenance mode bypassing may be ?

**Comment by kpodemski** [comment]
@jf-viguier 

> I don't know why the cookie name is parametrable ? it could be fixed

It is much safer this way.

> The cookie value should be generated to be non guessable, my be 15 c or more

That's the idea.

> The cookie value should be time limited

It will be you who creates the cookie, having a mechanism that deletes local cookie after some time after changing the option is an overkill IMHO.

> This user case could be adapted to the profiler and maintenance mode bypassing may be ?

In the future, probably, but let's treat it as a good first step.

**Comment by unlocomqx** [comment]
I added the generator buttons

CleanShot 2023-06-07 at 08 28 03@2x

I modified the `GeneratableInput` but in a backwards compatible way.

**Comment by unlocomqx** [comment]
Thank you @jf-viguier 

> I don't know why the cookie name is parametrable ? it could be fixed

Personally, I use the xdebug helper extension which can only toggle the cookies XDEBUG_SESSION and XDEBUG_PROFILE etc. So I would configure XDEBUG_SESSION as the cookie name, for example.
So it's more flexible that way.

**Comment by jf-viguier** [comment]
thanks @unlocomqx 
xdebug helper extension seems a very specific user cas for me.
The cookie name could be fixed and could be overrided by a dev in a conf file ?

**Comment by jf-viguier** [comment]
@kpodemski 
> I don't know why the cookie name is parametrable ? it could be fixed
> It is much safer this way.

I disagree, If the value is long enough, changing cookie name doesn't have to change.
The debug cookie settings has to be as simple as possible for a marchant.

**Comment by kpodemski** [comment]
@jf-viguier it's about copying and pasting the values to the browser, let's not overthink stuff

> The debug cookie settings has to be as simple as possible for a merchant.

remember that they don't need to put any value there and debug mode will work as before ;-)

**Review by kpodemski — APPROVED** [review]

**Review by boherm — APPROVED** [review]
LGTM like this @unlocomqx, thanks!

**Review by sallemiines — COMMENTED** [review]
Hello @unlocomqx 

Thanx for the PR ! 
- [Tests autos](https://github.com/sallemiines/ga.tests.ui.pr/actions/runs/5321297231) OK 
- I tried to check the PR  => The generator buttons are not functionnal , you can check the attached screen recorder 


https://github.com/PrestaShop/PrestaShop/assets/31663369/264f2348-651b-4ccf-ab2e-1564922a3dc5

Please check and feedback ! 
Thank you

**Comment by unlocomqx** [comment]
Hello @sallemiines 

Thanks for checking the PR

Try running the `make` command in the terminal to rebuild the scripts, because I modified a [typescript file](https://github.com/PrestaShop/PrestaShop/pull/32809/commits/034899eead7f1cd5e2218ec7873b802c5f8ac96e#diff-56265646b3776516117116f8d61c017b304c55d3d220b8aaebf2f04c795d67bf)

Thanks

**Comment by sallemiines** [comment]
@unlocomqx  Thank you for your feedback ! 

- Generator buttons works well 

Performance-•-develop-pr


But I noticed that when I enable debug mode and generate the **debug cookie name** and **debug cookie value**, the Debug Mode icon which is at the top of the page no longer exists 
Is That the expected behavior ? you can check the attached screen recorder 


https://github.com/PrestaShop/PrestaShop/assets/31663369/28cc47f2-19eb-4a07-b551-3723334dab68




And could you please provide me how could I check or use  the  debug cookie name and debug cookie value generated  ? 

Waiting for your feedback ! 

Thank you !

**Comment by MatShir** [comment]
It is the expected behavior. It seems that it is well working for now, know you need to edit the cookie to be on debug mode.

**Comment by unlocomqx** [comment]
Here's the easiest way to test the cookie

https://github.com/PrestaShop/PrestaShop/assets/793712/4c810c87-980c-40ab-b254-ce131d2119e2

**Comment by jf-viguier** [comment]
Thanks @unlocomqx for your patience ...

**Comment by sallemiines** [comment]
> Here's the easiest way to test the cookie
> 
>  CleanShot.2023-06-20.at.17.16.29.mp4


Thank you :rocket: ! LGTM QA :heavy_check_mark:

**Comment by kpodemski** [comment]
Thanks @unlocomqx 🎉

Yet another great contribution!

**Comment by unlocomqx** [comment]
Thank you, glad to contribute

**Comment by Hlavtox** [comment]
@unlocomqx Hi, could you please make another PR and add `PS_DEBUG_COOKIE_NAME` `PS_DEBUG_COOKIE_VALUE` to configuration during install?

Also, please add it to https://github.com/PrestaShop/autoupgrade/blob/dev/upgrade/sql/9.0.0.sql and link the PR to description. ;-) Thank you

**Comment by unlocomqx** [comment]
Hi @Hlavtox, here's the new PR
https://github.com/PrestaShop/autoupgrade/pull/660

**Comment by Hlavtox** [comment]
@unlocomqx Thank you! :-)

Now, please add it here and we are ready for v9 - https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/data/xml/configuration.xml 👍

**Comment by unlocomqx** [comment]
Okay, I added it, please check

https://github.com/PrestaShop/PrestaShop/pull/34929

**Pull request merged**